var/spack/repos/builtin/packages/scalasca/package.py

```diff
@@ -38,6 +38,9 @@ class Scalasca(Package):
     homepage = "http://www.scalasca.org"
     url = "http://apps.fz-juelich.de/scalasca/releases/scalasca/2.1/dist/scalasca-2.1.tar.gz"
 
+    version('2.3.1', 'a83ced912b9d2330004cb6b9cefa7585',
+            url='http://apps.fz-juelich.de/scalasca/releases/scalasca/2.3/dist/scalasca-2.3.1.tar.gz')
+
     version('2.2.2', '2bafce988b0522d18072f7771e491ab9',
             url='http://apps.fz-juelich.de/scalasca/releases/scalasca/2.2/dist/scalasca-2.2.2.tar.gz')
```

Comment:
I just rewrote half of the URL parsing logic, so I know how it works pretty well now. The way we substitute versions is by finding the version (`2.1`) in the package-level `url` and substituting all occurrences of that version with the new version (`2.3.1`). However, `scalasca` has a different URL scheme where the path contains the "major.minor" version in one place and the "major.minor.patch" version in another. Spack's default URL substitution logic can't handle this so you need to help it out a little.

The way I see it, you should rely on a single `url` at the top of the package if that is enough to get things working. If it doesn't work, you should use a `url_for_version` if the URL scheme is fairly predictable (like `scalsca`). If the URL is completely unpredictable (see `mfem`), then you should specify a `url` for each version manually.

For this package, it would look like:
```python
def url_for_version(self, version):
    url = "http://apps.fz-juelich.de/scalasca/releases/scalasca/{0}/dist/scalasca-{1}.tar.gz"
    return url.format(version.up_to(2), version)
```